unpublished/build-system-abstraction-njs.rst

```diff
@@ -0,0 +1,708 @@
+explicitly requested build-dependencies. This has two benefits:
+
+- It allows for a single installation run to build multiple packages
+  that have contradictory build-requirements. E.g. if package1
+  build-requires pbr==1.8.1, and package2 build-requires pbr==1.7.2,
+  then these cannot both be installed simultaneously into the global
+  environment -- which is a problem when the user requests ``pip
+  install package1 package2``. Or if the user already has pbr==1.8.1
+  installed in their global environment, and a package build-requires
+  pbr==1.7.2, then downgrading the user's version would be rather
+  rude.
+
+- It acts as a kind of public health measure to maximize the number of
+  packages that actually do declare accurate build-dependencies. We
+  can write all the strongly worded admonitions to package authors we
+  want, but if build frontends don't enforce isolation by default,
+  then we'll inevitably end up with lots of packages on PyPI that
+  build fine on the original author's machine and nowhere else, which
+  is a headache that no-one needs.
+
+However, there will also be situations where build-requirements are
+problematic in various ways. For example, a package author might
+accidentally leave off some crucial requirement despite our best
+efforts; or, a package might declare a build-requirement on `foo >=
+1.0` which worked great when 1.0 was the latest version, but now 1.1
+is out and it has a showstopper bug; or, the user might decide to
+build a package against numpy==1.7 -- overriding the package's
+preferred numpy==1.8 -- to guarantee that the resulting build will be
+compatible at the C ABI level with an older version of numpy (even if
+this means the resulting build is unsupported upstream). Therefore,
+build frontends SHOULD provide some mechanism for users to override
+the above defaults. For example, a build frontend could have a
+``--build-with-system-site-packages`` option that causes the
+``--system-site-packages`` option to be passed to
+virtualenv-or-equivalent when creating build environments, or a
+``--build-requirements-override=my-requirements.txt`` option that
+overrides the project's normal build-requirements.
+
+The general principle here is that we want to enforce hygiene on
+package *authors*, while still allowing *end-users* to open up the
+hood and apply duct tape when necessary.
+
+
+======================
+ Source distributions
+======================
+
+For now, we continue with the legacy sdist format which is mostly
+undefined, but basically comes down to: a file named
+``{NAME}-{VERSION}.{EXT}``, which unpacks into a buildable source tree
+called ``{NAME}-{VERSION}/``. Traditionally these have always
+contained "version 0" source trees; we now allow them to also contain
+version 1+ source trees.
+
+Integration frontends require that an sdist named
+``{NAME}-{VERSION}.{EXT}`` will generate a wheel named
+``{NAME}-{VERSION}-{COMPAT-INFO}.whl``.
+
```

Comment:
I do like the fact that you formally record the requirement on the metadata (name, version) embedded in the sdist filename here. I don't actually think that's recorded in a (finalised) spec anywhere else.